conn_pool.go

```diff
@@ -103,12 +108,50 @@ func (p *ConnPool) acquire(deadline *time.Time) (*Conn, error) {
 		return c, nil
 	}
 
+	// Set initial timeout/deadline value. If the method (acquire) happens to
+	// recursively call itself the deadline should retain its value.
+	if deadline == nil && p.acquireTimeout > 0 {
+		tmp := time.Now().Add(p.acquireTimeout)
+		deadline = &tmp
+	}
+
+	// Make sure the deadline (if it is) has not passed yet
+	if p.deadlinePassed(deadline) {
+		return nil, errors.New("Timeout: Acquire connection timeout")
+	}
+
+	// If there is a deadline then start a timeout timer
+	if deadline != nil {
+		timer := time.AfterFunc(deadline.Sub(time.Now()), func() {
+			p.cond.Signal()
+		})
+		defer timer.Stop()
+	}
+
 	// No connections are available, but we can create more
 	if len(p.allConnections) < p.maxConnections {
-		c, err := p.createConnection()
+		// Create a placeholder connection.
+		placeholderConn := &Conn{}
+		p.allConnections = append(p.allConnections, placeholderConn)
+		// Create a new connection.
+		// Carefull here: createConnectionUnlocked() removes the current lock,
+		// creates a connection and then locks it back.
+		c, err := p.createConnectionUnlocked()
+		// Take the placeholder out of the list of connections.
+		p.removeFromAllConnections(placeholderConn)
+		// Make sure create connection did not fail
 		if err != nil {
 			return nil, err
 		}
+		// If resetCount was updated since we started our connection, or
+		// there is no room in the list of allConnections
+		// (invalidateAcquired may remove our placeholder), try to re-acquire
+		// the connection.
+		if len(p.allConnections) >= p.maxConnections {
```

Comment:
I'll change this to "==". I replaced "==" with ">=" to catch a bug, and forgot to switch it back.
Should be harmless though